narwhal/executor/src/subscriber.rs

```diff
@@ -55,47 +52,52 @@ struct Fetcher<Network> {
     metrics: Arc<ExecutorMetrics>,
 }
 
-#[must_use]
-pub fn spawn_subscriber(
+pub fn spawn_subscriber<State: ExecutionState + Send + Sync + 'static>(
     name: PublicKey,
     network: oneshot::Receiver<P2pNetwork>,
     worker_cache: SharedWorkerCache,
     committee: Committee,
     rx_reconfigure: watch::Receiver<ReconfigureNotification>,
     rx_sequence: metered_channel::Receiver<CommittedSubDag>,
-    tx_notifier: metered_channel::Sender<(BatchIndex, Batch)>,
     metrics: Arc<ExecutorMetrics>,
     restored_consensus_output: Vec<CommittedSubDag>,
-) -> JoinHandle<()> {
-    spawn_monitored_task!(async move {
-        // This is ugly but has to be done this way for now
-        // Currently network incorporate both server and client side of RPC interface
-        // To construct server side we need to set up routes first, which requires starting Primary
-        // Some cleanup is needed
-        let network = network.await.expect("Failed to receive network");
-        info!("Starting subscriber");
-        let network = SubscriberNetworkImpl {
-            name,
-            worker_cache,
-            committee,
-            network,
-        };
-        let fetcher = Fetcher {
-            network,
-            metrics: metrics.clone(),
-        };
-        let subscriber = Subscriber {
-            rx_reconfigure,
-            rx_sequence,
-            metrics,
-            tx_notifier,
-            fetcher,
-        };
-        subscriber
-            .run(restored_consensus_output)
-            .await
-            .expect("Failed to run subscriber")
-    })
+    state: State,
+) -> Vec<JoinHandle<()>> {
+    // This is ugly but has to be done this way for now
+    // Currently network incorporate both server and client side of RPC interface
+    // To construct server side we need to set up routes first, which requires starting Primary
+    // Some cleanup is needed
+
+    let (tx_notifier, rx_notifier) =
+        metered_channel::channel(primary::CHANNEL_CAPACITY, &metrics.tx_notifier);
+
+    vec![
+        spawn_monitored_task!(async move { run_notify(state, rx_notifier).await }),
+        spawn_monitored_task!(async move {
```

Comment:
Do you mind putting all that `async` code into separate method, perhaps `Subscriber::start`?

And then this line can become just `spawn_monitored_task!(Subscriber::start(...))`

Reply:
The reason why we spawn 2 monitored tasks is so that the two can be done in parallel and not just concurrently. Do you know if there is a way to make it parallel with just one task? I didn't think there was.

Reply:
We should spawn two tasks, I am just asking to move contents of `async` block into separate method

Reply:
ah I see